### PR TITLE
DeviceSourceManager: Fix typing for DeviceType to narrow options in DeviceSourceManager

### PR DESCRIPTION
**Pull request by PolygonalSun**
This PR fixes an issue where the DeviceType/DeviceInput narrowing in DeviceSourceManager wasn't working properly.  With this change, the type of input used can be evaluated with Intellisense and prevent mixing and matching of inputs that _shouldn't_ belong together.

Examples:
DeviceSourceType -
image
DeviceSourceEvent -
image

InputManager (checking if deviceSource from onDeviceConnectedObservable is DeviceType.Mouse - 
image
Same File, evaluating event param in onInputChangedObservable -
image
Evaluating after PointerInput Mouse wheel checks -
image

In BabylonReactNative's App.tsx (demonstrating typing on getDeviceSource) -
image
Same File but with an added line to demonstrate getInput's rejection of invalid DeviceInputs -
image

### CODE REVIEW

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts [hunk 1]
```diff
@@ -49,40 +49,27 @@ export class DeviceSourceManager implements IDisposable, IObservableManager {
...
     /**
      * Default constructor
-     * @param engine Used to get canvas (if applicable)
+     * @param engine - Used to get canvas (if applicable)
      */
     constructor(engine: Engine) {
         const numberOfDeviceTypes = Object.keys(DeviceType).length / 2;
-        this._devices = new Array<Array<DeviceSource<DeviceType>>>(numberOfDeviceTypes);
-        this._firstDevice = new Array<number>(numberOfDeviceTypes);
+        this._devices = new Array(numberOfDeviceTypes);
```
Comment (ryantrem) [inline packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:71]: Why this change? Isn't this totally un-typesafe?

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts [hunk 2]
```diff
@@ -91,12 +78,18 @@ export class DeviceSourceManager implements IDisposable, IObservableManager {
...
-        this.onDeviceConnectedObservable = new Observable<DeviceSource<DeviceType>>((observer) => {
-            this.getDevices().forEach((device) => {
-                this.onDeviceConnectedObservable.notifyObserver(observer, device);
-            });
+        this.onDeviceConnectedObservable = new Observable((observer) => {
+            for (const devices of this._devices) {
+                if (devices) {
+                    for (const device of devices) {
+                        if (device) {
+                            this.onDeviceConnectedObservable.notifyObserver(observer, device as DeviceSourceType);
```
Comment (ryantrem) [inline packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:86]: I guess this cast is fall out from removing the type info for `_devices`?

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts [hunk 3]
```diff
@@ -125,12 +118,13 @@ export class DeviceSourceManager implements IDisposable, IObservableManager {
...
-     * @param deviceSource
+     * @param deviceSource - Source to add
+     // eslint-disable-next-line tsdoc/syntax
      * @hidden
      */
-    public _addDevice(deviceSource: DeviceSource<DeviceType>): void {
+    public _addDevice(deviceSource: DeviceSourceType): void {
         if (!this._devices[deviceSource.deviceType]) {
-            this._devices[deviceSource.deviceType] = new Array<DeviceSource<DeviceType>>();
+            this._devices[deviceSource.deviceType] = new Array();
```
Comment (ryantrem) [inline packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:127]: Similar question, seems like we've lost type safety here.

File: packages/dev/core/src/Inputs/scene.inputManager.ts [hunk 4]
```diff
@@ -865,47 +875,49 @@ export class InputManager {
         this._deviceSourceManager.onDeviceConnectedObservable.add((deviceSource) => {
             if (deviceSource.deviceType === DeviceType.Mouse) {
                 deviceSource.onInputChangedObservable.add((eventData) => {
-                    if (eventData.inputIndex === PointerInput.LeftClick || eventData.inputIndex === PointerInput.MiddleClick || eventData.inputIndex === PointerInput.RightClick) {
-                        const evt = eventData as IPointerEvent;
-                        if (attachDown && deviceSource.getInput(evt.inputIndex) === 1) {
-                            this._onPointerDown(evt);
-                        } else if (attachUp && deviceSource.getInput(evt.inputIndex) === 0) {
-                            this._onPointerUp(evt);
+                    if ((eventData.inputIndex === PointerInput.LeftClick || eventData.inputIndex === PointerInput.MiddleClick || eventData.inputIndex === PointerInput.RightClick)) {
```
Comment (ryantrem) [inline packages/dev/core/src/Inputs/scene.inputManager.ts:878]: Did this line get an extra set of parens?

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts (hunk 1 above)
Comment (PolygonalSun) [reply packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:71]: As far as I understand, Typescript should be able to deduce the array's object type from the initial declaration so `<Array<DeviceSource<DeviceType>>>` would be unnecessary.  Is this incorrect?

File: legacy/dist/preview release/what's new.md [hunk 5]
```diff
@@ -417,6 +417,7 @@
 - Fix spherical harmonics computation ([Meakk](https://github.com/Meakk))
 - Fix KTX and DDS loading with baked mipmaps ([Meakk](https://github.com/Meakk))
 - Fix text rendering speed when `TextWrapping.Ellipsis` is used ([carolhmj](https://github.com/carolhmj))
+- Fix issue where type narrowing wasn't working with DeviceType in DeviceSourceManager ([PolygonalSun](https://github.com/PolygonalSun))
```
Comment (bghgary) [inline legacy/dist/preview release/what's new.md:420]: Please remove this from what's new. We didn't ship an incorrect type narrowing in 4.2.

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts [hunk 6]
```diff
@@ -91,12 +78,18 @@ export class DeviceSourceManager implements IDisposable, IObservableManager {
         this._engine._deviceSourceManager._refCount++;
 
         // Observables
-        this.onDeviceConnectedObservable = new Observable<DeviceSource<DeviceType>>((observer) => {
-            this.getDevices().forEach((device) => {
-                this.onDeviceConnectedObservable.notifyObserver(observer, device);
-            });
+        this.onDeviceConnectedObservable = new Observable((observer) => {
+            for (const devices of this._devices) {
+                if (devices) {
```
Comment (bghgary) [inline packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:83]: Is it possible for this to be missing?

File: packages/dev/core/src/Events/deviceInputEvents.ts [hunk 7]
```diff
@@ -101,6 +103,11 @@ export interface IKeyboardEvent extends IUIEvent {
  */
 export interface IMouseEvent extends IUIEvent {
     // Properties
```
Comment (bghgary) [inline packages/dev/core/src/Events/deviceInputEvents.ts:105]: I would recommend removing these lines. It's confusing and unnecessary.
```suggestion
```

File: packages/dev/core/src/Inputs/scene.inputManager.ts [hunk 8]
```diff
@@ -500,6 +509,7 @@ export class InputManager {
         }
         this._deviceSourceManager = new DeviceSourceManager(engine);
 
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (bghgary) [inline packages/dev/core/src/Inputs/scene.inputManager.ts:512]: Why? If the variable is unused, just put an underscore in front of unused var to quiet.

File: packages/dev/core/src/Inputs/scene.inputManager.ts [hunk 9]
```diff
@@ -976,7 +988,8 @@ export class InputManager {
     }
 
     /**
-     * @param mesh
+     * @param mesh - Mesh to invalidate
+     // eslint-disable-next-line tsdoc/syntax
```
Comment (bghgary) [inline packages/dev/core/src/Inputs/scene.inputManager.ts:992]: Do we have to do this to all `@hidden` functions?

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts (hunk 1 above)
Comment (ryantrem) [reply packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:71]: Ah yea ok that makes sense.

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts (hunk 2 above)
Comment (PolygonalSun) [reply packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:86]: I believe that it's forcing us to cast because DeviceSource<DeviceType> isn't being seen as equal to DeviceSourceType

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts (hunk 3 above)
Comment (PolygonalSun) [reply packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:127]: This should be the same case as the initial instantiation for _devices

File: packages/dev/core/src/Events/deviceInputEvents.ts (hunk 7 above)
Comment (PolygonalSun) [reply packages/dev/core/src/Events/deviceInputEvents.ts:105]: Removed

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts (hunk 6 above)
Comment (PolygonalSun) [reply packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:83]: With gamepads I believe that this is possible

File: packages/dev/core/src/Inputs/scene.inputManager.ts (hunk 8 above)
Comment (PolygonalSun) [reply packages/dev/core/src/Inputs/scene.inputManager.ts:512]: Fixed

File: packages/dev/core/src/Inputs/scene.inputManager.ts (hunk 4 above)
Comment (PolygonalSun) [reply packages/dev/core/src/Inputs/scene.inputManager.ts:878]: Yeah, looks like it was a possible copy-paste error.  Removed parens

File: packages/dev/core/src/Inputs/scene.inputManager.ts (hunk 9 above)
Comment (PolygonalSun) [reply packages/dev/core/src/Inputs/scene.inputManager.ts:992]: Just asked in chat and we do not.  Removed

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts (hunk 6 above)
Comment (bghgary) [reply packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:83]: Can you confirm? It seems weird to have a map with null entries.

**Review by ryantrem — APPROVED** [review]

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts (hunk 6 above)
Comment (PolygonalSun) [reply packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:83]: I can confirm this.  The `_devices` object is not technically a map but just an array of an array.  When we first instantiate it, we just create an entry for each and every DeviceType.  This means that there could be devices that haven't initialized their arrays (eg. _devices[DeviceType.Keyboard]).  If all else fails, we _could_ convert it to a Map, which would remove the need to check if a DeviceSource array for a specific DeviceType exists.

Comment (bghgary) [reply packages/dev/core/src/DeviceInput/InputDevices/deviceSourceManager.ts:83]: Okay, sounds good.

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSource.ts [hunk 10]
```diff
@@ -7,11 +7,13 @@ import { IKeyboardEvent, IPointerEvent, IWheelEvent } from "../../Events/deviceI
...
-    T extends DeviceType.Mouse ? IWheelEvent | IPointerEvent :
-    T extends DeviceType.Touch ? IPointerEvent :
-    never;
+export type DeviceSourceEvent<T extends DeviceType> = T extends DeviceType.Keyboard
+    ? IKeyboardEvent
+    : T extends DeviceType.Mouse
+    ? IWheelEvent | IPointerEvent
+    : T extends DeviceType.Touch
+    ? IPointerEvent
+    : never;
```
Comment (bghgary) [inline packages/dev/core/src/DeviceInput/InputDevices/deviceSource.ts:16]: nit: This formatting is hard to read. Did autoformatting do this?

**Review by bghgary — APPROVED** [review]

File: packages/dev/core/src/Events/deviceInputEvents.ts [hunk 11]
```diff
@@ -221,7 +224,7 @@ export interface IMouseEvent extends IUIEvent {
  * Native friendly interface for PointerEvent Object
  */
 export interface IPointerEvent extends IMouseEvent {
-    // Properties
+    inputIndex: Exclude<PointerInput, PointerInput.Horizontal | PointerInput.Vertical | PointerInput.MouseWheelX | PointerInput.MouseWheelY | PointerInput.MouseWheelZ>;
```
Comment (bghgary) [inline packages/dev/core/src/Events/deviceInputEvents.ts:227]: nit: maybe add comment like IMouseEvent and no newline after (same for IWheelEvent)

File: packages/dev/core/src/DeviceInput/InputDevices/deviceSource.ts (hunk 10 above)
Comment (PolygonalSun) [reply packages/dev/core/src/DeviceInput/InputDevices/deviceSource.ts]: I won't clean this up in this PR but I'll handle it once everything's been ironed out with the new build process.

File: packages/dev/core/src/Events/deviceInputEvents.ts (hunk 11 above)
Comment (PolygonalSun) [reply packages/dev/core/src/Events/deviceInputEvents.ts:230]: Fixed
